### DISCUSSION TITLE
Update state on checkbox onChange

### DISCUSSION THREAD

**Discussion by nickcanarelli** [react-hook-form/react-hook-form#2894] [Q&A] [answered]
I am trying to add a class to a div when the checkbox is checked or remove the class if it's unchecked. I am using the multi-step form wizard example from react-hook-form using little-state-machine. How would I go about setting the state when it's checked/unchecked instead of onsubmit? 

I have the useForm({ mode: 'onChange', }); but that doesn't seem to make a difference.

Thanks in advance.

**Comment by bluebill1049** [comment]
codesandbox?

**Reply by nickcanarelli** [reply]
https://codesandbox.io/s/icy-dew-9un1c

**Comment by ekimkael** [comment, marked as answer]
First you have to `watch` that input and conditionally add `class`. Here is how did it.
But in my case I watch a field to affect another but it's the same process

```
{fields.map((passenger, index) => {
  const identity_type = watch(
    `passengers[${index}].identity_type`
  )
  return (
    <fieldset key={passenger.id} className="passengers">
      <div className="form-group">
        <label htmlFor={`passengers[${index}].identity_type`}>
          <FormattedMessage id="passengersStep.poi" />
        </label>
        <div className="input-group">
          <select
            className="custom-select"
            name={`passengers[${index}].identity_type`}
            id={`passengers[${index}].identity_type`}
            ref={register()}
          >
            <option defaultValue value="cni">
              Carte d'Identité
            </option>
            <option value="passeport">Passeport</option>
            <option value="schoolcard">Carte Scolaire</option>
            <option value="minor">Mineur/Enfant</option>
          </select>

          <input
            type="tel"
            className="form-control"
            name={`passengers[${index}].identity_value`}
            placeholder="119001234"
            ref={register()}
            disabled={identity_type === "minor" ? true : false}
          />
        </div>
        <ErrorMessage
          errors={errors}
          name={`passengers[${index}].identity_value`}
        >
          {({ message }) => (
            <small className="form-text text-danger">
              {message}
            </small>
          )}
        </ErrorMessage>
      </div>
    </fieldset>
  )
})}
```

**Reply by bluebill1049** [reply]
Thanks for helping!

**Reply by nickcanarelli** [reply]
@ekimkael you are a lifesaver! @bluebill1049 you've built quite a community - keep it up!

**Comment by nickcanarelli** [comment]
Just another quick question, when integrating with Yup validation - what would my approach be to require that at least 1 checkbox is checked within the form?

the returned data looks like:
```
"options":{
    "option1": {
        "id":"option1"
        "is_active": true
    }
    "option2": {
        "id":"option2"
        "is_active": false
    }
    ...
}

```

Right now i have: 
```
options: Yup.array()
    .of(
      Yup.object().shape({
        is_active: Yup.boolean().oneOf([true]),
      })
    )
    .required('Must choose at least one option.'),
```

but im getting the following error when submitting:
`options must be a 'array' type, but the final value was: 'null'`

**Reply by ekimkael** [reply]
First thing, do all fields have the same name?
Or what are you trying to do?

**Reply by nickcanarelli** [reply]
@ekimkael it's the same array of objects as the one in the CSB. I'm trying to add validation to it so that at least 1 checkbox needs to be checked. And i assume the validation would need to be set to the is_active key/value in the object to determine if it's checked or not?

I've updated the CSB to reflect validation, watch, etc...

https://codesandbox.io/s/icy-dew-9un1c

**Reply by ekimkael** [reply]
@nickcanarelli here is a CSB don't know if it's the right answer to your issue but check it and let me know
[CodeSanBox](https://codesandbox.io/s/checkboxes-and-radio-yup-rhf-example-3gju8)

**Reply by nickcanarelli** [reply]
@ekimkael Didnt seem to work. I'm just not sure how to validation dynamically created checkboxes.... any ideas?

**Reply by ekimkael** [reply]
Sorry I haven't try this for now.
But I think the main thing to understand/know is: are these fields have the same `name`? Are they intended to validate/add options to something?
`name` is the reference to validate the form. That's why I asked what you're trying to do

**Reply by nickcanarelli** [reply]
So the names are dynamically created via the following: 

``` 
{options.map((option) => {
    const optionKey = option.id;
    const name = `options[${optionKey}].is_active`;
    const watchActive = watch(`options[${optionKey}].is_active`);

    return (
        <Col lg={4} key={optionKey} className="d-block">
            <div
                className={`onboard__label ${
                    watchActive === true ? "is-checked" : ""
                }`}
            >
                {option.name}
                <input
                  type="hidden"
                  name={`options[${optionKey}].id`}
                  defaultValue={optionKey}
                  ref={register}
                />
                <input
                  type="checkbox"
                  ref={register}
                  name={name}
                  defaultChecked={option.is_active}
                />
            </div>
        </Col>
    );
})}
```

The names are not identical which is why i'm looking to validate dynamically created checkboxes. They're not intended to add options to something other than the overall state managed by little-state-machine. All i want to do is ensure that at least one checkbox out of the entire map is selected on submit, if not return error.

**Reply by ekimkael** [reply]
Okay, I'll see it tomorrow. It's already late on my timezone but i suggest you to look [FieldArray section](https://react-hook-form.com/advanced-usage#FieldArrays) especially the **register** tab

**Reply by nickcanarelli** [reply]
> Okay, I'll see it tomorrow. It's already late on my timezone but i suggest you to look [FieldArray section](https://react-hook-form.com/advanced-usage#FieldArrays) especially the **register** tab

So i've looked at that but i'm not sure how that would help given that I don't want every checkbox to be required. I only want **at least one** checkbox required on submit. If none are checked, return error. If one is selected, move on.

**Reply by bluebill1049** [reply]
This is a build in validation which you can get some idea form: https://codesandbox.io/s/checkbox-min-checked-5vocd

**Reply by bluebill1049** [reply]
I assume for the schema validation would be fair simpler as long as your schema will check one of the checkbox is checked.

**Reply by bluebill1049** [reply]
One trick I would do is testing `getValues()` aginst your schema and see if that's working or not.

**Reply by nickcanarelli** [reply]
@bluebill1049 right, i'd like to use schema validation as that would be more consistent throughout the form. Though i'm not sure how I would check to see what the Yup validation would look like for such a scenario.

**Reply by bluebill1049** [reply]
```
resolver: (data) => {
  console.log(data). // watch the value
  console.log(schema.validate(data)) // test it here

  return {
     values: {},
     errors: {}
  }
}
```

**Reply by nickcanarelli** [reply]
> ```
> resolver: (data) => {
>   console.log(data). // watch the value
>   console.log(schema.validate(data)) // test it here
> 
>   return {
>      values: {},
>      errors: {}
>   }
> }
> ```

I apologize im a bit lost here.. care to explain?

**Reply by bluebill1049** [reply]
```
useForm({
  resolver: (data) => {
  console.log(data). // watch the value
  console.log(schema.validate(data)) // test it here

  return {
     values: {},
     errors: {}
  }
}
})
```
